### PR TITLE
[Docs] Add multi-step pipeline routing example

### PR DESCRIPTION
**Pull request by xiaoyu-xyz**
## Summary
- add keyword signal and composite decision fragments for client-orchestrated multi-step routing
- document how step markers such as `__pipeline_step:summarize__` and `__pipeline_step:extract__` can pin each step to a model
- clarify that the client still orchestrates multiple requests while Semantic Router handles per-step model routing

## Validation
- `agent-ci-gate CHANGED_FILES="config/signal/keyword/multi-step-pipeline.yaml,config/decision/composite/multi-step-pipeline.yaml,website/docs/tutorials/decision/multi-step-pipeline.md,website/docs/tutorials/decision/overview.md,website/sidebars.ts"`

Refs #1815

### CODE REVIEW

File: config/decision/composite/multi-step-pipeline.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        operator: AND
+        conditions:
+          - type: keyword
+            name: pipeline_step_summarize
+          - operator: NOT
+            conditions:
+              - type: keyword
+                name: pipeline_step_extract
+      modelRefs:
+        - model: summarizer-model
```
Comment (rootfs) [inline config/decision/composite/multi-step-pipeline.yaml]: [risk:high] The model identifier `summarizer-model` looks like a placeholder; ensure this exact model name exists in operator configuration or document that operators must map this to a real model to avoid runtime routing failures.

File: config/decision/composite/multi-step-pipeline.yaml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        operator: AND
+        conditions:
+          - type: keyword
+            name: pipeline_step_extract
+          - operator: NOT
+            conditions:
+              - type: keyword
+                name: pipeline_step_summarize
+      modelRefs:
+        - model: insight-model
```
Comment (rootfs) [inline config/decision/composite/multi-step-pipeline.yaml]: [risk:high] The model identifier `insight-model` is unvalidated here; add an explicit note or validation step to prevent silent misconfiguration at deploy time.

File: config/signal/keyword/multi-step-pipeline.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+routing:
+  signals:
+    keywords:
+      - name: pipeline_step_summarize
+        operator: OR
+        keywords: ["__pipeline_step:summarize__", "__step1__"]
```
Comment (rootfs) [inline config/signal/keyword/multi-step-pipeline.yaml:6]: [risk:medium] The keywords `"__pipeline_step:summarize__"` and `"__step1__"` are lexically visible to models; recommend guidance on choosing markers that won't collide with user content and whether to escape or wrap them.

File: config/decision/composite/multi-step-pipeline.yaml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  decisions:
+    - name: pipeline_summarize_step
+      description: Pin the summarization stage of a client-orchestrated pipeline.
+      priority: 190
+      rules:
+        operator: AND
+        conditions:
+          - type: keyword
+            name: pipeline_step_summarize
+          - operator: NOT
```
Comment (rootfs) [inline config/decision/composite/multi-step-pipeline.yaml:11]: [risk:medium] The `NOT` guard prevents dual markers but behavior for requests containing both is ambiguous; add a test or explicit documented fallback explaining which rule wins.

File: website/docs/tutorials/decision/multi-step-pipeline.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
+# Multi-Step Pipeline Routing
+
+## Overview
+
+Use this pattern when a client already orchestrates a fixed multi-step workflow, but each step should be routed to a different model by Semantic Router.
+
+Semantic Router does not execute a full pipeline graph for the client. The client still sends one request per step. Each request carries a stable step marker, such as `__pipeline_step:summarize__` or `__pipeline_step:extract__`, and normal keyword signals map those markers to route decisions.
```
Comment (rootfs) [inline website/docs/tutorials/decision/multi-step-pipeline.md:7]: [risk:medium] Docs note that markers are model-visible; add an explicit example showing how to wrap markers in an app-controlled wrapper or instruct models to ignore them to avoid prompt-injection or leakage concerns.

File: website/sidebars.ts [hunk 6]
```diff
@@ -126,6 +126,7 @@ const sidebars: SidebarsConfig = {
             'tutorials/decision/or',
             'tutorials/decision/not',
             'tutorials/decision/composite',
+            'tutorials/decision/multi-step-pipeline',
```
Comment (rootfs) [inline website/sidebars.ts:129]: [risk:medium] New sidebar entry `tutorials/decision/multi-step-pipeline` must match the actual doc ID/path; verify the docs build resolves this path and update if the route differs to avoid broken navigation.

File: config/signal/keyword/multi-step-pipeline.yaml [hunk 7]
```diff
@@ -0,0 +1,11 @@
+routing:
+  signals:
+    keywords:
+      - name: pipeline_step_summarize
```
Comment (rootfs) [inline config/signal/keyword/multi-step-pipeline.yaml:4]: [risk:medium] Add unit tests for the `pipeline_step_summarize` signal to assert detection for both marker variants and case-insensitivity (`case_sensitive: false`).

File: config/decision/composite/multi-step-pipeline.yaml [hunk 8]
```diff
@@ -0,0 +1,32 @@
+routing:
+  decisions:
+    - name: pipeline_summarize_step
+      description: Pin the summarization stage of a client-orchestrated pipeline.
+      priority: 190
+      rules:
+        operator: AND
+        conditions:
+          - type: keyword
```
Comment (rootfs) [inline config/decision/composite/multi-step-pipeline.yaml:9]: [risk:medium] Add unit tests for the `pipeline_summarize_step` decision to ensure rules (keyword match + NOT guard) choose `summarizer-model` and that priority interacts as expected with other decisions.

File: config/decision/composite/multi-step-pipeline.yaml [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+      modelRefs:
+        - model: summarizer-model
+          use_reasoning: false
+    - name: pipeline_extract_step
+      description: Pin the extraction stage of a client-orchestrated pipeline.
+      priority: 190
+      rules:
+        operator: AND
+        conditions:
+          - type: keyword
```
Comment (rootfs) [inline config/decision/composite/multi-step-pipeline.yaml:24]: [risk:medium] Similarly add tests for `pipeline_extract_step` (keyword + NOT guard) to confirm `insight-model` selection and to catch ambiguous-marker edge cases.

**Review by rootfs — CHANGES_REQUESTED** [review]
Verdict
request_changes — docs + config are clear but two actionable issues: unvalidated model refs and missing tests/validation.

Risk
- Unknown model identifiers may break routing at deploy time (config/decision/composite/multi-step-pipeline.yaml:N16, N31).
- Step markers are model-visible and could leak or confuse prompts if misused (website/docs/tutorials/decision/multi-step-pipeline.md:N7, N93-N98).
- Docs/site integration risk: adding sidebar entry must match docs build expectations; a wrong path will fail the site build (website/sidebars.ts:N129 → website/docs/tutorials/decision/multi-step-pipeline.md:N37-N39).

Tests
- No tests added for new signals/decisions. Add unit tests verifying keyword detection and decision matching for:
  - config/signal/keyword/multi-step-pipeline.yaml:N4-N10
  - config/decision/composite/multi-step-pipeline.yaml:N9-N15, N24-N31
- Also add/verify website build test covers the new page and sidebar entry.

Style/Hygiene
- YAML fragments follow repo style and include guards (NOT) — good. Consider documenting example client request payload snippet showing where to place markers (docs: website/docs/tutorials/decision/multi-step-pipeline.md:N95-N97).
- Consider clarifying that model names are placeholders or require operator config.

Size
- Scope is small (5 files, focused docs + config). No need to split, but require the small follow-ups above before merging.

Tools used: NONE

**Comment by xiaoyu-xyz** [comment]
Thanks for the review. I pushed an update addressing the requested follow-ups:

- Replaced the placeholder `modelRefs` with reference-config model names (`qwen3-8b` and `qwen3-32b`) and documented that operators should replace them with models registered in their own deployment.
- Added a concrete client request payload showing the step marker inside an app-controlled wrapper, with guidance that markers are model-visible, should not come from user input, and should be ignored by the target model.
- Documented the ambiguous-marker behavior: if both step markers are present, the `NOT` guards prevent both pipeline decisions from matching and normal fallback routing applies.
- Added fragment-backed unit tests for keyword marker detection, alias detection, case-insensitive matching, per-step decision selection, priority behavior against a generic route, and the ambiguous-marker no-match case.

Local validation passed:

- `go test ./pkg/classification ./pkg/decision -run 'TestMultiStepPipeline' -count=1`
- `npm run build` in `website/`
- `make agent-ci-gate` with the PR changed files

**Comment by xiaoyu-xyz** [comment]
Hi @rootfs, I rebased this PR on the latest main and all checks are green now. The requested follow-ups were addressed in the latest commits. Could you take another look when you have a chance?

**Review by Xunzhuo — CHANGES_REQUESTED** [review]
The earlier code and documentation follow-ups are present, but this still has no accepted implementation issue. #1815 is needs-acceptance and describes a different runtime contract. Please link an accepted issue that owns this example and test scope before another review.

**Comment by FAUST-BENCHOU** [comment]
Thanks for the contribution. The example is clear, but I think this pattern is better handled at the application/workflow layer rather than encoded as prompt keywords. Semantic Router should focus on routing decisions based on request context and metadata instead of becoming a pipeline state coordinator. Closing this for now, but the workflow metadata based approach could be revisited in the future.
